tests/native_api_tests/test_api_model.cpp

```diff
@@ -351,37 +375,183 @@ TEST_CASE("API Model") {
+                CHECK(batch_node_result_id[1] == 4);
+                CHECK(batch_node_result_energized[1] == 0);
+                CHECK(batch_node_result_u[1] == doctest::Approx(0.0));
+                CHECK(batch_node_result_u_pu[1] == doctest::Approx(0.0));
+                CHECK(batch_node_result_u_angle[1] == doctest::Approx(0.0));
+            }
+        }
+    }
+
+    SUBCASE("Model update error optional id") {
+        std::vector<ID> const input_node_id{0};
+        std::vector<double> const input_node_u_rated{100.0};
+        Buffer input_node_buffer{PGM_def_input_node, 1};
+        input_node_buffer.set_nan();
+        input_node_buffer.set_value(PGM_def_input_node_id, input_node_id.data(), -1);
+        input_node_buffer.set_value(PGM_def_input_node_u_rated, input_node_u_rated.data(), -1);
+
+        std::vector<ID> const input_source_id{1};
+        std::vector<ID> const input_source_node{0};
+        std::vector<int8_t> const input_source_status{1};
+        std::vector<double> const input_source_u_ref{1.0};
+        std::vector<double> const input_source_sk{1000.0};
+        std::vector<double> const input_source_rx_ratio{0.0};
+        Buffer input_source_buffer{PGM_def_input_source, 1};
+        input_source_buffer.set_nan();
+        input_source_buffer.set_value(PGM_def_input_source_id, input_source_id.data(), -1);
+        input_source_buffer.set_value(PGM_def_input_source_node, input_source_node.data(), -1);
+        input_source_buffer.set_value(PGM_def_input_source_status, input_source_status.data(), -1);
+        input_source_buffer.set_value(PGM_def_input_source_u_ref, input_source_u_ref.data(), -1);
+        input_source_buffer.set_value(PGM_def_input_source_sk, input_source_sk.data(), -1);
+        input_source_buffer.set_value(PGM_def_input_source_rx_ratio, input_source_rx_ratio.data(), -1);
+
+        std::vector<ID> const input_sym_load_id{2};
+        std::vector<ID> const input_sym_load_node{0};
+        std::vector<int8_t> const input_sym_load_status{1};
+        std::vector<int8_t> const input_sym_load_type{2};
+        std::vector<double> const input_sym_load_p_specified{0.0};
+        std::vector<double> const input_sym_load_q_specified{500.0};
+        Buffer input_sym_load_buffer{PGM_def_input_sym_load, 1};
+        input_sym_load_buffer.set_nan();
+        input_sym_load_buffer.set_value(PGM_def_input_sym_load_id, input_sym_load_id.data(), -1);
+        input_sym_load_buffer.set_value(PGM_def_input_sym_load_node, input_sym_load_node.data(), -1);
+        input_sym_load_buffer.set_value(PGM_def_input_sym_load_status, input_sym_load_status.data(), -1);
+        input_sym_load_buffer.set_value(PGM_def_input_sym_load_type, input_sym_load_type.data(), -1);
+        input_sym_load_buffer.set_value(PGM_def_input_sym_load_p_specified, input_sym_load_p_specified.data(), -1);
+        input_sym_load_buffer.set_value(PGM_def_input_sym_load_q_specified, input_sym_load_q_specified.data(), -1);
```

Comment:
out of scope of this PR but the more i see it, the more i'm leaning towards making a utility for this. this is a very recurrent pattern haha. food for thought.